doc/sphinx/addendum/rewrite-rules.rst

```diff
@@ -0,0 +1,178 @@
+User-defined rewrite rules
+==========================
+
+.. warning::
+
+   The status of rewrite rules is highly experimental.
+
+   In particular, ill-typed rewrite rules will lead to mistyped expressions,
+   and manipulating these will most often result in anomalies.
+
+
+This section describes the extension of Coq's reduction mechanisms with user-defined rewrite rules,
+as a means to extend definitional equality. It should not be confused with the :ref:`rewrite tactic <rewritingexpressions>`
+or :ref:`setoid rewriting <generalizedrewriting>` which operate on propositional equality and other relations which are defined in Coq.
+
+The use of rewrite rules is guarded behind a flag, which can be enabled
+either by passing ``-allow-rewrite-rules`` to the
+Coq program or by turning the :flag:`Allow Rewrite Rules` flag on.
+
+.. flag:: Allow Rewrite Rules
+
+   This :term:`flag` enables the use of rewrite rules.
+   It is disabled by default and cannot be disabled once enabled.
+   The command-line flag ``-allow-rewrite-rules`` enables rewrite rules at
+   startup.
+
+   .. exn:: Rewrite rule declaration requires enabling the flag "Allow Rewrite Rules".
+      :undocumented:
+
+Symbols
+-----------------
+
+Rewrite rules operate on symbols, which are their own kind of constants.
+They stand in-between defined constants and axioms,
+in that they don't always reduce as defined constants do,
+but they may still reduce using the provided rules, unlike axioms.
+
+.. cmd:: {| Symbol | Symbols } {| {+ ( @assumpt ) } | @assumpt }
+   :name: Symbol; Symbols
+
+   As a command, `Symbol` binds an :n:`ident` to a :n:`type` as a symbol.
+
+   .. coqtop:: in
+
+      Set Allow Rewrite Rules.
+      Symbol pplus : nat -> nat -> nat.
+      Notation "a ++ b" := (pplus a b).
+
+Rewrite rules
+---------------
+
+  .. insertprodn rewrite_rule rewrite_rule
+
+  .. prodn::
+     rewrite_rule ::= {? @univ_decl %|- } @rw_pattern ==> @term
+
+Rewrite rules have two parts simply named left-hand side or pattern and right-hand side.
+When a rule is applied, the term is matched against the pattern,
+subterms aligned with pattern variables are collected
+and substituted into the right-hand side which is returned.
+
+.. cmd:: Rewrite {| Rule | Rules } @ident := {? %| } {+| @rewrite_rule }
+  :name: Rewrite Rule; Rewrite Rules
+
+   The command declares a named block of rewrite rules.
+
+  .. coqtop:: all
+
+     Rewrite Rule pplus_rewrite :=
+     | ?n ++ 0 ==> ?n
+     | ?n ++ S ?m ==> S (?n ++ ?m)
+     | 0 ++ ?n ==> ?n
+     | S ?n ++ ?m ==> S (?n ++ ?m).
+
+Pattern syntax
+--------------
+
+Patterns are a subset of terms which follow a more rigid syntax.
+It is easier to understand them as being a rigid head-pattern, inside of an elimination context:
+
+  .. prodn::
+     rw_head_pattern ::= @qualid {? @univ_annot }
+     | fun ({+ @name } {? : @rw_pattern_arg}) => @rw_pattern_arg
+     | forall ({+ @name } {? : @rw_pattern_arg}), @rw_pattern_arg
+     elim_context ::= []
```

Comment:
The `[]` are literal characters in this production?

Reply:
See the comment on `elim_context[rw_head_pattern]`